reporting/src/main/scala/caliban/reporting/ReportingDaemon.scala

```diff
@@ -0,0 +1,82 @@
+package caliban.reporting
+
+import zio.clock.Clock
+import zio.duration.{ durationInt, Duration }
+import zio.{ Has, UIO, ZIO, ZLayer, ZManaged }
+
+/**
+ * Manages schema reporting on start up, can be provided a set of ReportingRefs which will then be periodically
+ * sent to apollo
+ */
+trait ReportingDaemon {
+
+  /**
+   * Returns the list of refs that are being tracked by the daemon
+   */
+  def refs: UIO[List[SchemaReportingRef]]
+}
+
+object ReportingDaemon {
+
+  def layer(
+    ref: SchemaReportingRef,
+    rest: SchemaReportingRef*
+  ): ZLayer[Clock with Has[SchemaReporter], Nothing, Has[ReportingDaemon]] =
+    make(ref, rest: _*).toLayer
+
+  def make(
+    ref: SchemaReportingRef,
+    rest: SchemaReportingRef*
+  ): ZManaged[Clock with Has[SchemaReporter], Nothing, ReportingDaemon] =
+    ZManaged.suspend {
+      val schemaRefs = ref :: rest.toList
+      for {
+        reporter <- ZManaged.service[SchemaReporter]
+        _        <- ZIO
+                      .foreachPar_(schemaRefs) { ref =>
+                        def loop(withCoreSchema: Boolean): ZIO[Clock, Nothing, Unit] =
+                          reporter
+                            .report(ref, withCoreSchema)
+                            .foldM(
+                              {
+                                case ReportingError.SchemaError(_, _, message, _)  =>
+                                  ZIO.debug(s"Schema reporting failed for ${ref.graphRef}: $message")
+                                case ReportingError.ClientError(error)             =>
+                                  ZIO.debug(
+                                    s"Schema reporting for ${ref.graphRef} failed because of a client error ${error.getMessage}. This is likely a defect, will not retry"
+                                  )
+                                case ReportingError.RetryableError(innerThrowable) =>
+                                  ZIO.debug(
+                                    s"Schema reporting encountered an error: ${innerThrowable.getMessage}, retrying in 20 seconds"
+                                  ) *> loop(false).delay(20.seconds)
+                              },
+                              resp => loop(resp.withCoreSchema).delay(resp.in)
+                            )
+                        loop(false)
+                      }
+                      .forkManaged
```

Comment:
I'm debating if this or a "registration" based approach would be better, where we build the daemon and then register the schemas later. It takes a bit more complex bookkeeping though so I'm not sure if would actually be beneficial since most consumers would know their graphs upfront.

Reply:
I know there are a few users who don't use fixed schemas but build it dynamically, but this is probably a large minority.